### ISSUE TITLE
[Enhancement Proposal] Fix badly rendered shadows on common objects.

### ISSUE THREAD

**Issue by NikoVP** [diasurgical/devilutionX#2451]
This applies to all versions of Diablo

The graphics "bug" has been part of the game since release. Although the game is based on 3d models, the creators have used method that generates shadows for objects, from the sprites themselves by blackening and transforming the sprite to create a "shadow" and then combined it with the normal sprite. This method works some of the time, but in others its quite obvious that the result isn't true to what a properly rendered shadow is. Now finally after all these years we are at a point where correcting this is feasible. I have created a series of masks and new shadows that can be used to cutout the old and apply the new, leaving us with graphics the way they were supposed to be. Of course there many more cases in player, objects, monster sprites, but these are some of the most obvious cases. 

Just go and see all the chests, sarcophagus and weapons stands.

The graphics should contain correct shadows, not appearing out from the light source angle.

![image](https://user-images.githubusercontent.com/60973634/126851670-3ca50ab3-ed1e-4b5c-b755-1ea34506dbb0.png)

Proposal: I would gladly do all of it but I lack the programming skills. hence I am looking for collaborator here. A volunteer should use the masks and shadows provided in the file to modify the existing graphics. The provided graphic files are based on the PNG extracted version of the graphics, I believe that they should have the same pixel dimensions as the original CEL files. Further conversion of the files might be needed, to make them compatible with the current capabilities of the engine.

![image](https://user-images.githubusercontent.com/60973634/126852092-64db1e15-8bf2-4fc6-9d01-5207f273bd64.png)

Here is a zip file containing all the masks for all chests, cathedral sarcophagus and weapon stands.

[Objects Ready.zip](https://github.com/diasurgical/devilutionX/files/6871645/Objects.Ready.zip)

**Comment by AJenbo** [comment]
Co incidentally an issue with the render order of the original game caused this to be most hidden from the player as all sprites would effectively result in the following masking:
![image](https://user-images.githubusercontent.com/204594/126853177-09633ce6-eb86-4a93-b4a3-f4f41a8f4d81.png)

**Comment by AJenbo** [comment]
There are tow options for doing this.

1: Make the sprite type that uses SDL_Surface rather then CEL, on loading of the affected sprites covert the cel to an SDL_Surface and apply the mask.

2: Make a routine that can rencode the CEL data to change the transparent pixels, this is a bit tricky as this is done via the RLE compression.

I greatly favor option 1 and that would also make the game able to render sprites loaded from other format very easily.

**Comment by glebm** [comment]
I'm concerned about the increased memory usage of using `SDL_Surface` throughout. It'd be good if this was implemented as an option rather. Luckily, we now use the `CelSprite` wrapper class throghout, so that should be feasible.

**Comment by AJenbo** [comment]
@glebm what if we make sure to enable RLE compression for the SDL surfaces https://wiki.libsdl.org/SDL_SetSurfaceRLE ? I would think that should bring there size very close to that of CEL.

that being said i did intend for this to live along side the cel format and only be used when relevant

**Comment by glebm** [comment]
Yeah, RLE should be close to CEL. CL2 also has a "fill-N" command which SDL RLE does not, which might make a difference.

Not sure if SDL supports masked blits. If not, there is not much difference in implementing it for CEL vs CL2.

**Comment by FitzRoyX** [comment]
It's worth noting that sprites commonly make use of solid black, so the shadow cannot be accurately separated from them. You'd just be guessing where each begins and ends. There are 10k+ sprites. I'd consider this a huge timesink for a problem almost no one notices and has only an inconsistent solution at best.

**Comment by nickgavran** [comment]
> It's worth noting that sprites commonly make use of solid black, so the shadow cannot be accurately separated from them. 

Hey, can you provide an example of this, because I am very confident this is not true at all. If sprites "commonly" used solid black, it would mean that the trn/pal files would also translate the black shadow into other colors, and this does not occur in any part of the game. I scoured many sprites that look solid black, and the only ones to truly use a black color with the values '000000' (same as the shadow), is Tremain the Priest (which is an unused sprite) and Town Portal. 

2 sprites is hardly common.

**Comment by FitzRoyX** [comment]
Well, you're right that it's not common, so I take that part back. The sprite colors are so dark that it's hard to tell without manipulating the brightness higher. I was poking at the rogue boots when I made that assumption.

**Comment by NikoVP** [comment]
Nothing in this method affects sprite colours, It affects what parts of the image is displayed by effectively subtracting/removing all information that we don't need. Again it all depends how it's implemented. All I've done is to provide the 'needed parts for the repairwork' and point the problem.

**Comment by julealgon** [comment]
@nickgavran this is what I meant back in discord when I mentioned the shadows looked wrong to me:
| Vanilla | Proposed |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/461579/127050886-9e90a29b-1d4b-444d-a844-0fdfc09a9e39.png) | ![image](https://user-images.githubusercontent.com/461579/127051086-ee4b20a3-ef09-42ae-ac45-5ba2dac3b867.png) |

At the time, you said the left part of the shadow (in the pick above) looked fine, while I said it looked incorrect to me.

The proposal looks a lot more "correct" to me (even though I can't really explain it properly 😆 )

**Comment by julealgon** [comment]
> There are tow options for doing this.
> 
> 1: Make the sprite type that uses SDL_Surface rather then CEL, on loading of the affected sprites covert the cel to an SDL_Surface and apply the mask.
> 
> 2: Make a routine that can rencode the CEL data to change the transparent pixels, this is a bit tricky as this is done via the RLE compression.
> 
> I greatly favor option 1 and that would also make the game able to render sprites loaded from other format very easily.

Or option 3: make this a part of the installation process (which currently doesn't exist 🤣 )

**Comment by nickgavran** [comment]
> @nickgavran this is what I meant back in discord when I mentioned the shadows looked wrong to me:
> 
> Vanilla	Proposed
> ![image](https://user-images.githubusercontent.com/461579/127050886-9e90a29b-1d4b-444d-a844-0fdfc09a9e39.png)	![image](https://user-images.githubusercontent.com/461579/127051086-ee4b20a3-ef09-42ae-ac45-5ba2dac3b867.png)
> At the time, you said the left part of the shadow (in the pick above) looked fine, while I said it looked incorrect to me.
> 
> The proposal looks a lot more "correct" to me (even though I can't really explain it properly 😆 )

The proposal looks better, as the shadow is actually coming from the correct angle there. The shadow we currently have looks fine to me when the sarcophagus is OPEN, maybe I didn't clarify this on discord. Oh well

**Comment by glebm** [comment]
These days the best option would probably be to apply the masks during CLX conversion.
This can be done in the engine at load time, and in https://github.com/diasurgical/devilutionx-mpq-tools/ for low-RAM builds.

**Comment by donatj** [comment]
If this is “fixed” can it be optional as the original shadows are perhaps part of the charm for some people?

**Comment by AJenbo** [comment]
> If this is “fixed” can it be optional as the original shadows are perhaps part of the charm for some people?

The original cut them because of overdraw, so it's not an original charm, see https://github.com/diasurgical/devilutionX/issues/2451#issuecomment-885979314
